canary/webrtc-c/src/media-server-storage/kvsWebRTCClientMaster.cpp

```diff
@@ -0,0 +1,367 @@
+    UINT64 t1;
+    BOOL initialized = FALSE;
+
+    SET_INSTRUMENTED_ALLOCATORS();
+    UINT32 logLevel = setLogLevel();
+
+
+    t1 = GETTIME() / HUNDREDS_OF_NANOS_IN_A_MILLISECOND;
+    std::string filePath = "../toConsumer.txt";
+    remove(filePath.c_str());
+
+    std::thread canaryDurationThread;
+
+    Aws::InitAPI(options);
+    CHK_STATUS(canaryConfig.init(argc, argv));
+    canaryConfig.isStorage = true;	
+    CHK_STATUS(Canary::Cloudwatch::init(&canaryConfig));
+    canaryConfig.print();
+    SET_LOGGER_LOG_LEVEL(canaryConfig.logLevel.value);
+    DLOGD("[Canary] Canary init time: %d [ms]", (GETTIME() / HUNDREDS_OF_NANOS_IN_A_MILLISECOND) - t1);
+
+#ifndef _WIN32
+    signal(SIGINT, sigintHandler);
+#endif
+
+#ifdef IOT_CORE_ENABLE_CREDENTIALS
+    CHK_ERR((pChannelName = getenv(IOT_CORE_THING_NAME)) != NULL, STATUS_INVALID_OPERATION, "AWS_IOT_CORE_THING_NAME must be set");
+#else
+    pChannelName = (PCHAR) canaryConfig.channelName.value.c_str();
+#endif
+
+    CHK_STATUS(createSampleConfiguration(pChannelName, SIGNALING_CHANNEL_ROLE_TYPE_MASTER, TRUE, TRUE, logLevel, &pSampleConfiguration));
+
+    // Set the audio and video handlers
+    pSampleConfiguration->audioSource = sendAudioPackets;
+    pSampleConfiguration->videoSource = sendVideoPackets;
+    pSampleConfiguration->receiveAudioVideoSource = sampleReceiveAudioVideoFrame;
+
+    // Set sample to use storage mode
+    pSampleConfiguration->channelInfo.useMediaStorage = TRUE;
+
+    // Initialize storage disconnected timestamp
+    pSampleConfiguration->storageDisconnectedTime = 0;
+
+
+#ifdef ENABLE_DATA_CHANNEL
+    pSampleConfiguration->onDataChannel = onDataChannel;
+#endif
+    pSampleConfiguration->mediaType = SAMPLE_STREAMING_AUDIO_VIDEO;
+    DLOGI("[KVS Master] Finished setting handlers");
+
+    // Check if the samples are present
+    DLOGI("[KVS Master] Checking sample video frame availability....");
+    CHK_STATUS(readFrameFromDisk(NULL, &frameSize, "./assets/h264SampleFrames/frame-0001.h264"));
+    DLOGI("[KVS Master] Checked sample video frame availability....available");
+
+    CHK_STATUS(readFrameFromDisk(NULL, &frameSize, "./assets/opusSampleFrames/sample-001.opus"));
+    DLOGI("[KVS Master] Checked sample audio frame availability....available");
+
+    // Initialize KVS WebRTC. This must be done before anything else, and must only be done once.
+    CHK_STATUS(initKvsWebRtc());
+    DLOGI("[KVS Master] KVS WebRTC initialization completed successfully");
+    initialized = TRUE;
+
+    CHK_STATUS(initSignaling(pSampleConfiguration, SAMPLE_MASTER_CLIENT_ID));
+    DLOGI("[KVS Master] Channel %s set up done ", pChannelName);
+
+    if(canaryConfig.duration.value != 0) {
+        DLOGD("[Canary] Scheduling canary duration for %lu seconds", canaryConfig.duration.value / HUNDREDS_OF_NANOS_IN_A_SECOND);
+        canaryDurationThread = std::thread(shceduleShutdown, canaryConfig.duration.value, pSampleConfiguration);
+        canaryDurationThread.detach();
```

Comment:
No need to detach. We should join().

Reply:
Joining here would cause us to wait here, but we need to get to the `sessionCleanupWait(...)` just below and let the `canaryDurationThread` exist in the background.

Reply:
We dont join here. We join after `sessionCleanUpWait()`, it should not hang at that point, correct?

Reply:
We remain within `sessionCleanUpWait` in a while loop for the remainder of the run though, so we wouldn't ever reach that code. I moved the this code over into `sessionCleanUpWait` just above the while loop, but still need to use detach. We could do away with the whole shutdown thread and just use that while loop to check for duration reached, how does that sound?

Reply:
Yes. Lets do that. No need of a new thread if `sessionCleanUpWait` is going to run through and through anyways.